File: src/ladym/operations/supersedes.py
"""supersedes pointer chain (SPEC §2.6).

Retired memories stay in the store (so lineage / audit is preserved) but are flagged so:

* ``UPDATE`` consolidation: a NEW merged memory is created, the OLD target is marked with
  ``metadata.superseded_by = <new.id>`` and a ``supersedes`` edge old→new is written. The
  old target is no longer returned by tier-1 recall, but tier-2 walks the edge to surface
  the newest version.
* ``DELETE`` consolidation: the target is marked ``metadata.superseded = true`` and has no
  successor. Tier-1 hides it; tier-2 has no edge to follow.
"""

from __future__ import annotations

import time

from ..schema import Edge, Memory
from ..storage.store import SQLiteStore


def is_retired(mem: Memory | None) -> bool:
    """True iff ``mem`` was retired by an UPDATE or DELETE consolidation pass.

    ``None`` (memory gone) is treated as not-retired so callers can keep applying the
    filter inline without a separate None-guard.
    """
    if mem is None:
        return False
    meta = mem.metadata or {}
    return bool(meta.get("superseded_by") or meta.get("superseded"))


def retire(store: SQLiteStore, old: Memory, *, new_id: str | None = None) -> None:
    """Retire ``old``.

    * ``new_id`` given  → UPDATE chain: write ``superseded_by`` + a ``supersedes`` edge
      so :func:`latest_in_chain` can walk to the successor.
    * ``new_id`` is None → DELETE retirement: set ``superseded=true`` and write no edge.

    Outgoing still-valid edges of ``old`` are closed (``valid_to=now``) so the graph does
    not leak through a retired node.
    """
    now = time.time()
    old.metadata = {**(old.metadata or {}), "superseded_at": now}
    if new_id:
        old.metadata["superseded_by"] = new_id
    else:
        old.metadata["superseded"] = True
    # close still-open outgoing edges first so neighbours do not keep flowing through a
    # retired node. Done BEFORE writing the new ``supersedes`` edge so we don't close
    # the very edge we are about to create (recall relies on it staying open).
    for e in store.neighbors(old.id):
        if e.valid_to is None:
            e.valid_to = now
            store.put_edge(e)
    if new_id:
        store.put_edge(
            Edge(src_id=old.id, relation="supersedes", dst_id=new_id, valid_from=now)
        )
    store.put_memory(old)


def latest_in_chain(store: SQLiteStore, mem_id: str) -> str:
    """Walk ``supersedes`` edges forward and return the id of the newest version.

    Stops at the head (a memory with no outgoing ``supersedes`` edge) and guards against
    cycles by tracking the visited set.
    """
    seen: set[str] = set()
    cur = mem_id
    while cur not in seen:
        seen.add(cur)
        nxt = [
            e
            for e in store.neighbors(cur, relation="supersedes")
            if e.src_id == cur
        ]
        if not nxt:
            return cur
        cur = nxt[0].dst_id
    return cur

File: operations/supersedes.go
package operations

import (
	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

// IsRetired reports whether mem was retired by an UPDATE or DELETE
// consolidation pass. A nil memory is treated as not-retired.
func IsRetired(mem *schema.Memory) bool {
	if mem == nil {
		return false
	}
	return mem.MetaString("superseded_by") != "" || mem.MetaBool("superseded")
}

// Retire retires old. When newID is non-empty it writes a supersedes edge
// old→new (UPDATE chain); otherwise it sets superseded=true (DELETE). Outgoing
// still-valid edges of old are closed so the graph does not leak through a
// retired node.
func Retire(store *storage.SQLiteStore, old *schema.Memory, newID string) error {
	now := schema.Now()
	if old.Metadata == nil {
		old.Metadata = map[string]any{}
	}
	old.Metadata["superseded_at"] = now
	if newID != "" {
		old.Metadata["superseded_by"] = newID
	} else {
		old.Metadata["superseded"] = true
	}
	neighbors, err := store.Neighbors(old.ID, "")
	if err != nil {
		return err
	}
	for _, e := range neighbors {
		if e.ValidTo == nil {
			e.ValidTo = &now
			if err := store.PutEdge(e); err != nil {
				return err
			}
		}
	}
	if newID != "" {
		e := schema.NewEdge(old.ID, "supersedes", newID)
		e.ValidFrom = now
		if err := store.PutEdge(e); err != nil {
			return err
		}
	}
	return store.PutMemory(old, nil)
}

// LatestInChain walks supersedes edges forward and returns the newest version id.
func LatestInChain(store *storage.SQLiteStore, memID string) (string, error) {
	seen := map[string]bool{}
	cur := memID
	for !seen[cur] {
		seen[cur] = true
		neighbors, err := store.Neighbors(cur, "supersedes")
		if err != nil {
			return cur, err
		}
		var best *schema.Edge
		for _, e := range neighbors {
			if e.SrcID != cur {
				continue
			}
			if best == nil || e.ValidFrom > best.ValidFrom {
				best = e
			}
		}
		if best == nil {
			return cur, nil
		}
		cur = best.DstID
	}
	return cur, nil
}
